improver/nowcasting/utilities.py

```diff
@@ -103,6 +101,130 @@ def process(self, radar_data, coverage):
         return cube_list.merge_cube()
 
 
+class FillRadarHoles(BasePlugin):
+    """Fill in small "no data" regions in the radar composite by interpolating
+    in log rainrate space."""
+
+    @staticmethod
+    def _rr_to_log_rr(data):
+        """Convert a masked array of rainrates in mm/h into log space.  This
+        does not preserve values below 0.001; however, since the radar encodes
+        trace rain rates with a value of 0.03 mm/h, this should not have any
+        effect on "real" data.
+        """
+        min_rr_mmh = 0.001
+        result = np.where(data > min_rr_mmh, np.log10(data), np.nan)
+        return np.ma.MaskedArray(result, mask=data.mask)
+
+    @staticmethod
+    def _log_rr_to_rr(data):
+        """Convert a masked array of log rainrates into mm/h"""
+        result = np.where(np.isfinite(data), np.power(10, data), 0.0)
+        return np.ma.MaskedArray(result, mask=data.mask)
+
+    @staticmethod
+    def _find_and_interpolate_speckle(log_rr):
+        """Identify and interpolate "speckle" points, where "speckle" is defined
+        as areas of "no data" that are small enough to fill by interpolation
+        without affecting data integrity.  We would not wish to interpolate large
+        areas as this may gives false confidence in "no precipitation", where in
+        fact precipitation exists in the "no data" region.
+
+        The constants for speckle identification ("p_masked" and "r_speckle")
+        have been empirically tuned for UK radar data.  With the constants
+        as set, this method will flag "holes" of up to 24 pixels in size.
+        The interpolation radius has been chosen to match these constants,
+        as the smallest radius that ensures there will always be valid data in
+        the neighbourhood (25 pixels) over which averaging is performed.
+
+        Masked pixels near the border of the input data array will not be
+        interpolated.
+
+        Args:
+            log_rr (numpy.ma.MaskedArray):
+                Masked array of rainrates in log10(mm/h)
+
+        Returns:
+            numpy.ma.MaskedArray:
+                Masked array of interpolated rainrates in log10(mm/h)
+        """
+        # proportion of masked neighbours below which a pixel is considered to
+        # be isolated "speckle", which can be filled in by interpolation
+        p_masked = 0.3
+        # radius of neighbourhood over which to search for masked neighbours
+        r_speckle = 4
+        # radius of neighbourhood from which to calculate interpolated values
+        r_interp = 2
+
+        interpolated_points = np.ma.MaskedArray(
+            np.full_like(log_rr, 0), mask=np.full_like(log_rr, True)
+        )
+
+        for y in range(r_speckle, log_rr.shape[0] - r_speckle - 1, 1):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable name "y" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)